cdc/metadata/metadata.go

```diff
@@ -0,0 +1,217 @@
+	SchedStopped
+)
+
+// ScheduledOwner is for owner and processor schedule.
+type ScheduledOwner struct {
+	*ChangefeedInfo
+	ChangefeedProgress
+	Capture *CaptureInfo
+	State   SchedState
+}
+
+// ScheduledProcessor is for owner and processor schedule.
+type ScheduledProcessor struct {
+	*ChangefeedInfo
+	ChangefeedProgress
+	Capture *CaptureInfo
+	State   SchedState
+}
+
+// ChangefeedSchedule is used to query changefeed schedule information.
+type ChangefeedSchedule struct {
+	Owner     ScheduledOwner
+	Processor []ScheduledProcessor
+}
+
+// CaptureObservation is for observing and updating metadata on a CAPTURE instance.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type CaptureObservation interface {
+	// CaptureInfo tells the caller who am I.
+	CaptureInfo() *CaptureInfo
+
+	// GetChangefeed queries one changefeed or all changefeeds.
+	GetChangefeeds(cf ...model.ChangeFeedID) ([]*ChangefeedInfo, error)
+
+	// GetCaptures queries one capture or all captures.
+	GetCaptures(cp ...string) ([]*CaptureInfo, error)
+
+	// Heartbeat tells the metadata storage I'm still alive.
+	Heartbeat(context.Context) error
+
+	// TakeControl blocks until becomes controller or gets canceled.
+	TakeControl() (ControllerObservation, error)
+
+	// Advance advances some changefeed progresses.
+	Advance(cfs []*ChangefeedInfo, progresses []*ChangefeedProgress) error
+
+	// Fetch the latest changefeed owner list from metadata.
+	RefreshOwners() <-chan []ScheduledOwner
+
+	// When an owner exits, inform the metadata storage.
+	PostOwnerStopped(cf *ChangefeedInfo) error
+
+	// Fetch the latest changefeed processors from metadata.
+	RefreshProcessors() <-chan []ScheduledProcessor
+
+	// When a processor exits, inform the metadata storage.
+	PostProcessorStopped(cf *ChangefeedInfo) error
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
+	// CreateChangefeed creates a changefeed.
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
+
+	// RemoveChangefeed removes a changefeed, will auto stop owner and processors.
+	RemoveChangefeed(cf *ChangefeedInfo) error
+
+	// Fetch the latest capture list in the TiCDC cluster.
+	RefreshCaptures() <-chan []*CaptureInfo
+
+	// Schedule a changefeed owner to a given target, or stop it if target is nil.
+	// Notes:
+	//   * the target capture can fetch the event by `RefreshOwners`.
+	//   * select `done` to wait the old owner in stopping to be resolved.
+	SetOwner(cf *ChangefeedInfo, target *CaptureInfo) (done <-chan struct{}, _ error)
+
+	// Schedule some captures as workers to a given changefeed.
+	// Notes:
+	//   * target captures can fetch the event by `RefreshProcessors`.
+	//   * select `done` to wait all processors in stopping to be resolved.
+	SetProcessors(cf *ChangefeedInfo, workers []*CaptureInfo) (done <-chan struct{}, _ error)
```

Comment:
Why does the Controller need to monitor the state of the processor?

Reply:
`done` is for being uniform with `SetOwner`. We can just print a log after processors in stopping get resolved.